### ISSUE TITLE
difference between BERT and GPT

### ISSUE THREAD

**Issue by hjq133** [huggingface/optimum#1247]
when I test bert-base-uncased and GPT2 with the benchmark_bettertransformer_training_minimal.py,
I find that GPT2 is accelerated well with optimum, but BERT-base-uncased does not.
here is my results:
GPT2: 
eager peak mem: 13291 MB
BT peak mem: 6803 MB

BERT-Base-uncase:
eager peak mem: 13632 MB
BT peak mem: 13627 MB

is optimum only work well for decoder-only architecture ?

**Comment by fxmarty** [comment]
Hi @hjq133, before, only decoder models were dispatching to flash attention during training. I decided by default to drop the attention/padding mask for all models in https://github.com/huggingface/optimum/pull/1250, which allows to effectively dispatch on flash attention / memory-efficient attention.

Could you try again on optimum 1.11?

You can read more about it here: https://huggingface.co/docs/optimum/main/en/bettertransformer/overview#quickstart

Let me know if anything can be improved!

**Comment by hjq133** [comment]
@fxmarty Hi
Thanks for your recommendation.
I just update my optimum to 1.11 and it works on my bert-uncased model now!
I get a speed up ratio 37%, and mem saving 79% on my A100 machine 

and it should only work on A100, right? 
I just test it on my V100 and there is no improve.

**Comment by fxmarty** [comment]
@hjq133 Thank you. You can use the decorator [`torch.backends.cuda.sdp_kernel`](https://pytorch.org/docs/master/backends.html#torch.backends.cuda.sdp_kernel) to check whether PyTorch is able to dispatch to flash attention or memory-efficient attention:

```python
with torch.backends.cuda.sdp_kernel(enable_flash=True, enable_math=False, enable_mem_efficient=True):
    # your inference/training code here
```

If you run into an error like "no kernel available", it means PyTorch just falls back to a C++ implementation of the attention which is not expected to be better than the eager pytorch in python. I will add it to the documentation, it is quite useful.

Maybe if you try on pytorch nightly version you can have a broader support compared to 2.0.1.
